src/fieldAnalyser/fieldAnalyser.ts

```diff
@@ -0,0 +1,138 @@
+import {BatchUpdateDocuments, DocumentBuilder} from '..';
+
+type fieldTypeDict = {[key in FieldTypes]?: number};
+type typeOccurence = [fieldType: string, occurence: number];
+export class FieldAnalyser {
+  private static fieldTypePrecedence = ['DOUBLE', 'STRING'];
+  private inconsitentFields: Record<string, FieldTypes[]> = {};
+
+  public constructor(private platformClient: PlatformClient) {}
+  public async createMissingFieldsFromBatch(batch: BatchUpdateDocuments) {
+    const alreadyCreatedFields = await this.listAllFieldsFromOrg();
+    const missingFieldsFromOrg = this.getMissingFieldsFromOrg(
+      batch,
+      alreadyCreatedFields
+    );
+    const fieldsToCreate = this.getFieldsToCreate(missingFieldsFromOrg);
+    return this.createFields(fieldsToCreate);
+  }
+
+  private async listAllFieldsFromOrg(): Promise<string[]> {
+    // TODO: CDX-836
+    throw new Error('Method not implemented.');
+  }
+
+  private createFields(_fields: Record<string, FieldTypes>) {
+    // TODO: CDX-835
+    throw new Error('Method not implemented.');
+  }
+
+  private getMissingFieldsFromOrg(
+    batch: BatchUpdateDocuments,
+    alreadyCreatedFields: string[]
+  ): Record<string, fieldTypeDict> {
+    const missingFieldsFromOrg: Record<string, fieldTypeDict> = {};
+
+    batch.addOrUpdate.flatMap((doc: DocumentBuilder) => {
+      const documentMetadata = Object.entries({...doc.build().metadata});
+
+      for (const [metadataKey, metadataValue] of documentMetadata) {
+        if (alreadyCreatedFields.includes(metadataKey)) {
+          continue;
+        }
+        const metadataType = this.getValue(metadataValue);
+        if (missingFieldsFromOrg[metadataKey]) {
+          // Possible metadata inconsitency
+          let fieldType = missingFieldsFromOrg[metadataKey][metadataType];
+          missingFieldsFromOrg[metadataKey][metadataType] = fieldType
+            ? ++fieldType
+            : 1;
+        } else {
+          missingFieldsFromOrg[metadataKey] = {[metadataType]: 1};
+        }
+      }
+    });
+
+    return missingFieldsFromOrg;
+  }
+
+  private getFieldsToCreate(
+    missingFieldsFromOrg: Record<string, fieldTypeDict>
+  ): Record<string, FieldTypes> {
+    const fieldsToCreate: Record<string, FieldTypes> = {};
+
+    Object.entries(missingFieldsFromOrg).map(([fieldName, fieldTypeDict]) => {
+      this.storePossibleTypeInconsistencies(fieldName, fieldTypeDict);
+      const fieldType = this.getMostProbableType(fieldTypeDict);
+      fieldsToCreate[fieldName] = fieldType;
+    });
+
+    this.warnAboutInconsistentFields();
+
+    return fieldsToCreate;
+  }
+
+  private storePossibleTypeInconsistencies(
+    fieldName: string,
+    fieldTypeDict: fieldTypeDict
+  ) {
+    const typePossibilities = Object.keys(fieldTypeDict) as FieldTypes[];
+    if (typePossibilities.length > 1) {
+      this.inconsitentFields[fieldName] = typePossibilities;
+    }
+  }
+
+  private getMostProbableType(field: fieldTypeDict): FieldTypes {
```

Comment:
Since you rely on the order of `fieldTypeDict` to do comparison and sorting, a `Map` seems more appropriate than a simple object.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map

Reply:
good point.